src/operator/z1.rs

```diff
@@ -1,11 +1,159 @@
+        &self.output
+    }
+
+    /// Connect `input` stream to the input of the `Z1` operator.
+    pub fn connect(self, input: &Stream<Circuit<P>, D>) {
+        let Self {
+            feedback,
+            output,
+            export,
+        } = self;
+        let circuit = output.circuit().clone();
+
+        feedback.connect_with_preference(input, OwnershipPreference::STRONGLY_PREFER_OWNED);
+        circuit
+            .cache()
+            .insert(DelayedId::new(input.local_node_id()), output);
+        circuit
+            .cache()
+            .insert(ExportId::new(input.local_node_id()), export);
+    }
+}
+
+/// Like [`FeedbackConnector`] but specialized for [`Z1Nested`] feedback
+/// operator.
+///
+/// Use this API instead of the low-level [`circuit::add_feedback`] API to
+/// create feedback loops with `Z1Nested` operator.  In addition to being more
+/// concise, this API takes advantage of [caching](`crate::circuit::cache`).
+pub struct DelayedNestedFeedback<P, D> {
+    feedback: FeedbackConnector<Circuit<P>, D, D, Z1Nested<D>>,
+    output: Stream<Circuit<P>, D>,
+}
+
+impl<P, D> DelayedNestedFeedback<P, D>
+where
+    P: Clone + 'static,
+    D: Clone + 'static,
+{
+    /// Create a feedback loop with `Z1` operator.  Use [`Self::connect`] to
+    /// close the loop.
+    pub fn new(circuit: &Circuit<P>, zero: D) -> Self {
+        let (output, feedback) = circuit.add_feedback(Z1Nested::new(zero));
+        Self { feedback, output }
+    }
+
+    /// Output stream of the `Z1Nested` operator.
+    pub fn stream(&self) -> &Stream<Circuit<P>, D> {
+        &self.output
+    }
+
+    /// Connect `input` stream to the input of the `Z1Nested` operator.
+    pub fn connect(self, input: &Stream<Circuit<P>, D>) {
+        let Self { feedback, output } = self;
+        let circuit = output.circuit().clone();
+
+        feedback.connect_with_preference(input, OwnershipPreference::STRONGLY_PREFER_OWNED);
+        circuit
+            .cache()
+            .insert(NestedDelayedId::new(input.local_node_id()), output);
+    }
+}
+
+impl<P, D> Stream<Circuit<P>, D> {
+    /// Applies [`Z1`] operator to `self`.
+    pub fn delay(&self) -> Stream<Circuit<P>, D>
+    where
+        P: Clone + 'static,
+        D: Clone + HasZero + 'static,
+    {
+        if let Some(delayed) = self
+            .circuit()
+            .cache()
+            .get(&DelayedId::new(self.local_node_id()))
+        {
+            return delayed.clone();
+        }
+
+        let delayed = self.circuit().add_unary_operator(Z1::new(D::zero()), self);
+        self.circuit()
+            .cache()
+            .insert(DelayedId::new(self.local_node_id()), delayed.clone());
+
+        delayed
+    }
+
+    /// Applies [`Z1Nested`] operator to `self`.
+    pub fn nested_delay(&self) -> Stream<Circuit<P>, D>
+    where
+        P: Clone + 'static,
+        D: Clone + HasZero + 'static,
+    {
+        if let Some(delayed) = self
```

Comment:
also this pattern is very common: if (lookup) return; create; cache; return.
But there is common code between lookup and cache; return. 
I wonder whether that can be factored using lambdas similar to the computeIfAbsent Java API https://docs.oracle.com/javase/8/docs/api/java/util/HashMap.html#computeIfAbsent-K-java.util.function.Function-